### PR TITLE
Add single cookie consent API

### PR DESCRIPTION
**Pull request by andysellick**
## What / why
Adds code for the [single consent cookie API](https://github.com/alphagov/consent-api) to allow cookie preferences to be shared across GOV.UK.

Supersedes https://github.com/alphagov/govuk_publishing_components/pull/3829

### How this works
The single consent API should take control of the setting of cookies relating to cookie consent. Fortunately it sets exactly the same cookies as we do now:

- `cookies_preferences_set` indicating that a user preference has been made
- `cookies_policy` showing the preferences chosen (set by default to no consent)

### Testing scenarios
There are two main things to test - when the consent API is enabled, and when it is not. When it is not enabled (by default) everything should work as normal. This means:

- a users first visit should set the `cookies_policy` to reject all but essential cookies, and no other cookies
- when they choose a preference (accept/reject/custom), this should be reflected in the `cookies_policy` cookie, and the `cookies_preferences_set` cookie should also be set
- the page at `/help/cookies` should reflect the contents of the `cookies_policy` cookie, and allow it to be changed

When the consent API is enabled, the behaviour should be similar but the consent API code should set these cookies, not our code. Additionally:

- no cookies are set until the user makes a choice
- when the user chooses a cookie preference (accept/reject/custom), the consent API code makes an XMLHttpRequest to the consent API
- if this is successful, it sets a `gov_singleconsent_uid` cookie, containing the users randomly generated anonymous ID
- if this fails (e.g. timeout), this cookie is not set and the consent preferences are all set to `reject`
- on any following pages, the consent API code will not attempt to reconnect to the API, and leave user consent as rejected

### Testing locally

Check out this branch and run `yarn install` to get the single consent API code.

To make testing easier and keep the gem compatible with applications that do not need the single consent API, it is disabled by default. It is enabled by setting `window.GOVUK.useSingleConsentApi = true`. For the purposes of testing this can be set in your local `static`.

You'll need to run `frontend` (it renders the cookies page) running locally through docker with a local `static`, both pointed at this branch of the components gem. You'll need to modify `config/initializers/content_security_policy.rb` in `frontend` so that the CSP allows JS to make requests to the consent API. The staging URL has changed during development, is currently `https://gds-single-consent-staging.app/`.

```
GovukContentSecurityPolicy.configure do |policy|
  policy.connect_src(*policy.connect_src, "<staging url>")
end
```

You can test the scenario of arriving at GOV.UK having set cookie preferences on another site by either:

- using a URL of the form: `https://www.gov.uk?gov_singleconsent_uid=<id>`. You can get the value for `id` by from the `gov_singleconsent_uid` cookie (after it has been successfully set)
- accept cookies and let the consent api set the `gov_singleconsent_uid` cookie, then delete all other cookies (this wouldn't happen in real life, but it's an easy way of simulating the consent api already being initialised)

To test link decoration going from GOV.UK to a participating site you can use the consent API staging site https://gds-single-consent-staging.app/ (this is why the staging URL is both the API interface and an actual website). Do the following:

- modify a page in `frontend` to have a link to the staging site
- in the components gem code, open `node_modules/govuk-single-consent/dist/singleconsent.iife.js` and modify the `function (origins)` code around line 357 to include the following line before setting `addUIDtoCrossOriginLinks`:

```
origins.push('https://gds-single-consent-staging.app') // or whatever the staging URL is
```

- click on the link
- the link should be automatically decorated with a URL parameter, will immediately disappear when the page has loaded, so you can check it's worked by comparing the cookie settings on the staging site with what you set on your local site. You can also scroll to the bottom of the staging site and click on the `Cookies` link to change your preferences, then go back to your local site to confirm that changing preferences works both ways.

### Cumulative Layout Shift impact

Broadly - no impact. I suspect this is mostly due to recent changes to hide the banner by default and when JS is disabled.

All tests run locally to account for differences in local and production, however testable scores on production are the same. Testing carried out locally on frontend, homepage.

**Desktop**

Without consent api | With consent api
-- | --
accept all cookies - **0** | accept all cookies - **0**
reject all cookies - **0** | reject all cookies - **0**
no cookie preference made - **0.31** | no cookie preference made - **0.31**
(not available) | with only a valid govuk_consent_id cookie (to force the consent api call) - **0**

**Mobile**

Without consent api | With consent api
-- | --
accept all cookies - **0.003** | accept all cookies - **0.003**
reject all cookies - **0.003** | reject all cookies - **0.003**
no cookie preference made - **0.486** | no cookie preference made - **0.486**
(not available) | with only a valid govuk_consent_id cookie (to force the consent api call) - **0.003**

## Visual Changes
None.

Trello card: https://trello.com/c/dmlTAzKB/140-implement-single-consent-api

### CODE REVIEW

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (AshGDS) [inline app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:41]: To reduce the layout shift, is there a way that the 'Your preferences have been set / Hide this message' part of the banner HTML can just not be shown at all when the consent API exists? Currently there's a split second where you can see the message before it disappears.

I found it a bit jarring for it to appear for a split second then disappear, and when I first saw it I thought it was a bug with the code. I'm a bit worried that when user's see that, they will think that it's an important bit of information that's disappearing before they can read it.  It also feels like it could be flagged as an accessibility issue, though I haven't looked into this.

**Review by AshGDS — CHANGES_REQUESTED** [review]
Thanks for doing this, I appreciate that it was a massive task to undertake 🎉 .

Overall the code looks good to me, though there are a few things I think we should figure out which I've left comments about.

It's still not working in IE11, but I've messaged Guilhem in our group DM about this.

Also - has there been a discussion on what happens when the user has two different consents set on two different `.gov.uk` domains? I'm still worried policy wise that we'll change the policy to say setting their cookies in one place will set it everywhere, when technically that's not the case. It only sets it in the same place if you click a link/to from `www.gov.uk` that goes to/from another approved origin. Right now, if you click a link it looks like the consent you have on the referrer i.e. `www.gov.uk` will then override any previous consent you set on that other domain when you click a link through to it. Which may cause issues if someone has a consent ID on `www.campaigns.gov.uk` and `www.gov.uk`. 

One issue is that they may want different settings for each domain. Maybe we need to provide an option for the user to be able to disable single consent themselves? 

Another issue is that their consent on a domain could change without them realising. For example if they go to `www.campaigns.gov.uk` via Google on Monday and set consent there, they will have a consent `uid` for `www.campaigns.gov.uk`. Then on Tuesday they go to `www.gov.uk` via Google. As cookies aren't shared across our domains, they will be treated as a brand new consent user and get another `uid`. Now they have two `uid` and there is no single consent  across these domains, but they may think they only have one `uid` and therefore changing it on one domain should change it on the other. I'm not sure this is something we can fix in this PR. I think I'm flagging it here so that it's logged somewhere that i'm worried the cookie policy won't reflect that this sort of scenario is possible to our users.

Overall, nice work 👍

File: app/assets/javascripts/govuk_publishing_components/components/cookie-banner.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/components/cookie-banner.js:90]: I think as it's already a call to another function I think I'll leave it as-is for now, rather than add more lines of code here, if that's okay.

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js]: Because I wasn't paying attention when I wrote this, apparently! 😄 Good spot, have corrected. I suspect I've done this in a few places..

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:70]: Unfortunately this single call to `setPreferences` with only the second parameter is outnumbered by the two calls to it in `cookie-banner.js` with only the first parameter. I think if it were switched around it would be more confusing.

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:17]: Yeah, I've been wondering about this. It started off as a small change and it's snowballed a bit. I think this is what I'd like to do:

- take the list of `domains` out of `load-analytics` and put it somewhere else, like a whole new file
- include the `loadExtraDomains` function with that
- leave the rest of the file as-is

Does that sound sensible?

File: app/assets/javascripts/govuk_publishing_components/components/cookie-banner.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/components/cookie-banner.js:90]: Cool, no worries

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:70]: No worries - I guess maybe the paramaters could be an object instead, but appreciate that I'm probably just introducing more complexity for something that doesn't really matter!

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:17]: Sounds good - regarding a new file, I'm guessing this would be another JS script to load into the DOM as well? This is because I believe `loadAnalytics` always loads before `application.js`, so we might be stuck trying to solve this as you would probably need these domains to exist before `loadAnalytics` is called.

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js]: I think I've massively over-complicated this, and only actually need to refer to the original `window` variable. I was thinking it would be good to have a single place where this was handled (in the consent library) and then refer to that, but it's just gotten way too complex. Have updated.

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:1]: Can you clarify the steps involved here? Are you talking about previously having given consent, then withdrawing it on the consent page?

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: Good shout. I think CLS is going to be an important part of this work. I'll investigate.

File: spec/javascripts/govuk_publishing_components/lib/cookie-settings-spec.js
Comment (maxgds) [inline spec/javascripts/govuk_publishing_components/lib/cookie-settings-spec.js:203]: Is it necessary to set and unset this before/after each test when (I believe) the value isn't manipulated by the tests?

Comment (maxgds) [reply spec/javascripts/govuk_publishing_components/lib/cookie-settings-spec.js:203]: I was more wondering if you could set it once for all (although there's only a couple) and unset at the end, rather than before/after each, but I guess it could fall foul of things running in parallel. No big deal either way.

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: I'm not sure I can reproduce the situation you're talking about, but I have identified a CLS problem when you've set preferences on another site and haven't been to GOV.UK before, which I think I've now fixed. Going to re-do the CLS testing again and add the numbers to the PR description.

File: spec/javascripts/govuk_publishing_components/lib/single-consent-functions-spec.js
Comment (andysellick) [reply spec/javascripts/govuk_publishing_components/lib/single-consent-functions-spec.js:202]: I think wider concerns like this are generally covered by other tests i.e. if something that big were missing an error would have already occurred e.g. removing `loadAnalytics` causes loads of unrelated tests to fail.

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:1]: @andysellick Yep, it exists on the live site currently so you haven't introduced it, but if you've already accepted usage cookies, and then you reject them on `/help/cookies`, our analytics tracking is still running after they've hit 'Save changes'. For example, if you right click on a footer link after setting usage cookies to false on `/help/cookies`, our tracking is still running.

File: spec/javascripts/govuk_publishing_components/lib/single-consent-functions-spec.js
Comment (AshGDS) [reply spec/javascripts/govuk_publishing_components/lib/single-consent-functions-spec.js:202]: @andysellick No worries

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:1]: Oh that's a good point. I'll start a new card to investigate this issue, thanks.

Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:1]: Card created: https://trello.com/c/vAI4kvX5/795-tracking-is-still-active-on-cookie-settings-page-after-rejecting-consent

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: Numbers added. Seems to be no impact on CLS, which I think is mainly due to the recent no-JS changes to the banner 👍

File: spec/javascripts/govuk_publishing_components/lib/cookie-settings-spec.js
Comment (andysellick) [reply spec/javascripts/govuk_publishing_components/lib/cookie-settings-spec.js:203]: I think I'll leave it as it is, if that's okay.

**Review by AshGDS — COMMENTED** [review]
Changes look good to me, let me know when you need another review 👍

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: Also, what is the context behind no longer showing the "Cookies accepted" message? Or is this unintentional?

**Comment by maxgds** [comment]
> One issue is that they may want different settings for each domain. Maybe we need to provide an option for the user to be able to disable single consent themselves?

I believe the intent is for all sites opting in to be using the same buckets for consent types and it's all or nothing. Although I do share your concerns when we raised this it was pushed back on as a non-technical question. I think Jonathan Nicolls was involved to some level around product decisions on this piece so might we worth chatting to him. Suffice to say, though, there aren't plans to have opt-outs beyond opting out of cookies totally.

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: Interesting - that shouldn't happen. I've just tried it by accepting cookies and it behaves as expected. Are you still able to reproduce?

Comment (maxgds) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: For what it's worth I can reproduce - haven't pulled the branch for a couple of days.

Comment (maxgds) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: https://github.com/alphagov/govuk_publishing_components/assets/31649453/f0015818-93ba-43d5-a93f-7a39854c6d2a

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:17]: Have come up with a solution, hopefully is sensible and understandable, see newest commit.

Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:17]: @andysellick Nice, looks good 👍 Do you think our tests should ensure the `production` values always stay the same in that object, or is that overkill? I guess there's a risk that the production values change without us knowing and that affects analytics/cookie consent.

Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:17]: It's a good thought, but I think it might be overkill. Two possible things I think you could be talking about here...

First, that the production details would be somehow lost during the `loadExtraDomains` function. This is semi-tested, in that the test first gets the hard coded domains, then appends the test data onto it, then checks that the output matches this. I guess we could hard code the expected production values (like you said) for `name` into the test, and check that, except that regardless of the environment, if we're passing extra values, the intent is that these extra values are used, not the defaults - so if the original production values are lost somehow, it shouldn't matter (unless we pass extra values on production, but that'd be strange).

Second, that someone makes a code change that alters the production values. I don't think we normally check that variables like this remain the same, and I'd hope that any accidental change to these values would be picked up in review (think you meant the first thing, but just covering all the options).

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: @andysellick Yep I still have it on the latest version of the branch.

Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:35]: @andysellick It's caused by:

`app/assets/javascripts/govuk_publishing_components/components/cookie-banner.js`, line 20:

```JS
window.addEventListener('cookie-consent', this.$module.hideCookieMessage)
```

That code will hide the cookie banner when the `cookie-consent` event is triggered.

This `cookie-consent` event is then being triggered by `app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js` line 33:

```JS
        if (consents && consents.usage) {
          window.GOVUK.triggerEvent(window, 'cookie-consent')
        } 
```

I think you can just remove the `cookie-consent` event listener that hides the cookie banner to fix it. What is it the context for having that listener?

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (AshGDS) [inline app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:34]: Forgot to mention this. Here you're deleting the `essential` cookie but not reintroducing it into the cookie object when the user submits the form.

Saving cookies isn't working properly in IE11 on this branch - it saves sucessfully, but then when I go to the homepage, the cookie settings I have in `document.cookie` are different to the ones that I saved.

Could this be because `essential` is missing from the cookie when the form is saved, causing something odd?

**Review by AshGDS — CHANGES_REQUESTED** [review]
@andysellick just adding something I forgot to add in the initial review

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js:17]: No worries - I was referring to the second thing 👍

File: app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js
Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/single-consent-functions.js]: Nice, seems to work for me now 👍

File: app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js
Comment (andysellick) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:34]: Is this a problem with this branch, or more generally? I don't think I introduced this change. Would it be alright to log this as an issue and for it to not block this change?

Comment (AshGDS) [reply app/assets/javascripts/govuk_publishing_components/lib/cookie-settings.js:34]: No worries - I've just tried it again and I think it's actually a weird IE11 caching thing. If I update my cookies, and go back to the homepage, the homepage gives me the old cookie state unless I add a cachebust query string to the URL. So don't worry about it

**Review by AshGDS — APPROVED** [review]
Nice work 👍

**Pull request merged**
